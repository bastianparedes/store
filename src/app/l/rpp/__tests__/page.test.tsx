/* eslint-disable sonarjs/no-duplicate-string */
import React from 'react';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';
import Page from '../page';

describe('Page', () => {
  it('renders Page component', () => {
    const { container } = render(<Page />);

    expect(container).toMatchSnapshot();
    expect(screen.getByLabelText(/Nombre del producto:/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Descripción:/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Cantidad disponible:/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Precio:/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Imagen:/i)).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Enviar/i })).toBeInTheDocument();
  });

  it('handles form input changes', () => {
    render(<Page />);

    fireEvent.change(screen.getByLabelText(/Nombre del producto:/i), {
      target: { value: 'Producto de prueba' }
    });
    fireEvent.change(screen.getByLabelText(/Descripción:/i), {
      target: { value: 'Descripción de prueba' }
    });
    fireEvent.change(screen.getByLabelText(/Cantidad disponible:/i), {
      target: { value: '5' }
    });
    fireEvent.change(screen.getByLabelText(/Precio:/i), {
      target: { value: '50' }
    });

    const file = new File(['(⌐□_□)'], 'test-file.png', { type: 'image/png' });
    fireEvent.change(screen.getByLabelText(/Imagen:/i), {
      target: { files: [file] }
    });

    expect((screen.getByLabelText(/Nombre del producto:/i) as any).value).toBe(
      'Producto de prueba'
    );
    expect((screen.getByLabelText(/Descripción:/i) as any).value).toBe(
      'Descripción de prueba'
    );
    expect((screen.getByLabelText(/Cantidad disponible:/i) as any).value).toBe(
      '5'
    );
    expect((screen.getByLabelText(/Precio:/i) as any).value).toBe('50');
  });

  it('handles submit success', async () => {
    global.fetch = jest
      .fn()
      .mockReturnValueOnce({ json: async () => ({ success: true }), ok: true });
    global.alert = jest.fn();
    Object.defineProperty(window, 'location', {
      value: {
        assign: jest.fn(),
        hash: {
          endsWith: jest.fn(),
          includes: jest.fn()
        }
      },
      writable: true
    });

    render(<Page />);

    fireEvent.change(screen.getByLabelText(/Nombre del producto:/i), {
      target: { value: 'Producto de prueba' }
    });
    fireEvent.change(screen.getByLabelText(/Descripción:/i), {
      target: { value: 'Descripción de prueba' }
    });
    fireEvent.change(screen.getByLabelText(/Cantidad disponible:/i), {
      target: { value: '5' }
    });
    fireEvent.change(screen.getByLabelText(/Precio:/i), {
      target: { value: '50' }
    });

    const file = new File(['(⌐□_□)'], 'test-file.png', { type: 'image/png' });
    fireEvent.change(screen.getByLabelText(/Imagen:/i), {
      target: { files: [file] }
    });

    const submitButton = screen.getByRole('button', { name: 'Enviar' });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalled();
      expect(window.location.pathname).toBe('/l/plp');
    });
  });

  it('handles submit fail', async () => {
    global.fetch = jest.fn().mockReturnValueOnce({
      json: async () => ({ success: false }),
      ok: true
    });
    global.alert = jest.fn();
    Object.defineProperty(window, 'location', {
      value: {
        assign: jest.fn(),
        hash: {
          endsWith: jest.fn(),
          includes: jest.fn()
        }
      },
      writable: true
    });

    window.location.pathname = '/l/rpp';

    render(<Page />);

    fireEvent.change(screen.getByLabelText(/Nombre del producto:/i), {
      target: { value: 'Producto de prueba' }
    });
    fireEvent.change(screen.getByLabelText(/Descripción:/i), {
      target: { value: 'Descripción de prueba' }
    });
    fireEvent.change(screen.getByLabelText(/Cantidad disponible:/i), {
      target: { value: '5' }
    });
    fireEvent.change(screen.getByLabelText(/Precio:/i), {
      target: { value: '50' }
    });

    const file = new File(['(⌐□_□)'], 'test-file.png', { type: 'image/png' });
    fireEvent.change(screen.getByLabelText(/Imagen:/i), {
      target: { files: [file] }
    });

    const submitButton = screen.getByRole('button', { name: 'Enviar' });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(global.fetch).toHaveBeenCalled();
      expect(window.location.pathname).toBe('/l/rpp');
    });
  });
});
