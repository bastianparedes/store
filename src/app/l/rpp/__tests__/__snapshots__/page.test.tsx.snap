// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders Page component 1`] = `
<div>
  <div
    class="container"
  >
    <form>
      <label
        for="name"
      >
        Nombre del producto:
        <input
          id="name"
          name="name"
          type="text"
          value=""
        />
      </label>
      <label
        for="description"
      >
        Descripci√≥n:
        <input
          id="description"
          name="description"
          type="text"
          value=""
        />
      </label>
      <label
        for="quantity"
      >
        Cantidad disponible:
        <input
          id="quantity"
          max="100"
          min="0"
          name="quantity"
          type="number"
          value="1"
        />
      </label>
      <label
        for="price"
      >
        Precio:
        <input
          id="price"
          max="999999"
          min="0"
          name="price"
          type="number"
          value="1000"
        />
      </label>
      <label
        for="image"
      >
        Imagen:
        <input
          id="image"
          name="picture"
          type="file"
        />
      </label>
      <button
        type="submit"
      >
        Enviar
      </button>
    </form>
  </div>
</div>
`;
